plugins/aws/sts_provisioner.go

```diff
@@ -77,3 +90,273 @@ func (p STSProvisioner) Deprovision(ctx context.Context, in sdk.DeprovisionInput
+	if awsConfig.HasSSOStartURL() {
+		return nil, fmt.Errorf(unsupportedMessage, "SSO Authentication", "https://github.com/1Password/shell-plugins/issues/210")
+	}
+
+	if awsConfig.HasWebIdentity() {
+		return nil, fmt.Errorf(unsupportedMessage, "Web Identity Authentication", "https://github.com/1Password/shell-plugins/issues/209")
+
+	}
+
+	if awsConfig.HasCredentialProcess() {
+		return nil, fmt.Errorf(unsupportedMessage, "Credential Process Authentication", "https://github.com/1Password/shell-plugins/issues/213")
+
+	}
+
+	if awsConfig.HasSourceProfile() {
+		return nil, fmt.Errorf(unsupportedMessage, "Sourcing profiles", "https://github.com/1Password/shell-plugins/issues/212")
+
+	}
+
+	if awsConfig.HasRole() {
+		return p.providerFactory.NewAssumeRoleProvider(awsConfig), nil
+	}
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
+		return p.providerFactory.NewMFASessionTokenProvider(awsConfig), nil
+	}
+
+	return p.providerFactory.NewAccessKeysProvider(), nil
+}
+
+type STSProviderFactory interface {
+	SetState(in sdk.ProvisionInput, out *sdk.ProvisionOutput)
+	NewAssumeRoleProvider(awsConfig *confighelpers.Config) aws.CredentialsProvider
+	NewMFASessionTokenProvider(awsConfig *confighelpers.Config) aws.CredentialsProvider
+	NewAccessKeysProvider() aws.CredentialsProvider
+}
+
+// CacheProviderFactory 's functions return AWS providers that are also handling reading and writing from shell plugin's encrypted cache
+type CacheProviderFactory struct {
+	InCache    sdk.CacheState
+	OutCache   sdk.CacheOperations
+	ItemFields map[sdk.FieldName]string
+}
+
+func (m *CacheProviderFactory) SetState(in sdk.ProvisionInput, out *sdk.ProvisionOutput) {
+	m.InCache = in.Cache
+	m.OutCache = out.Cache
+	m.ItemFields = in.ItemFields
+}
+
+func (m CacheProviderFactory) NewAssumeRoleProvider(awsConfig *confighelpers.Config) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN, m.ItemFields[fieldname.AccessKeyID])
+	if m.InCache.Has(roleCacheKey) {
+		return NewStsCacheProvider(roleCacheKey, m.InCache)
+	}
+
+	cacheWriter := NewSTSCacheWriter(roleCacheKey, m.OutCache)
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
+		return initAssumeRoleProvider(awsConfig, getSTSClient(awsConfig.Region, m.NewMFASessionTokenProvider(awsConfig)), cacheWriter)
+	}
+
+	return initAssumeRoleProvider(awsConfig, getSTSClient(awsConfig.Region, m.NewAccessKeysProvider()), cacheWriter)
+}
+
+func (m CacheProviderFactory) NewMFASessionTokenProvider(awsConfig *confighelpers.Config) aws.CredentialsProvider {
+	mfaCacheKey := getMfaCacheKey(m.ItemFields[fieldname.AccessKeyID])
+	if m.InCache.Has(mfaCacheKey) {
+		return NewStsCacheProvider(mfaCacheKey, m.InCache)
+	}
+
+	if awsConfig.NonChainedGetSessionTokenDuration == 0 {
+		awsConfig.NonChainedGetSessionTokenDuration = 900 * time.Second // default to minimum duration of 15 minutes for security
```

Comment:
super nit: we have `time.Minute` and I am not a fan of "for security" as it does not really explain any reasoning.
```suggestion
		// Set a low session TTL by default. 
		// 15 minutes should be enough to complete most operations.
		awsConfig.NonChainedGetSessionTokenDuration = 15 * time.Minute
```

Reply:
Sounds good, changed to minutes and removed the "for security"